File: Python/ByRound/1933/1933_F_Turtle_Mission_Robot_and_the_Earthquake.py
# Problem: CF 1933 F - Turtle Mission: Robot and the Earthquake
# https://codeforces.com/contest/1933/problem/F

"""
Algorithm: BFS with state tracking
Time Complexity: O(n * m)
Space Complexity: O(n * m)

The problem models a robot moving on a grid where columns move cyclically upwards over time.
We use BFS to explore the shortest path from (0,0) to (n-1,m-1).
The state includes position (i,j) and time t.
At each step, we can move up, down, or right, but movement depends on the rock positions
which shift cyclically with time.
"""

from sys import stdin

input = lambda: stdin.buffer.readline().decode().strip()
from collections import deque

for _ in range(int(input())):
    n, m = map(int, input().split())
    g = [list(map(int, input().split())) for _ in range(n)]
    ans = inf = 1 << 60
    vis = [[0] * m for _ in range(n)]

    # BFS queue stores (row, col, time)
    bfs = deque([(0, 0, 0)])
    vis[0][0] = 1
    while bfs:
        i, j, t = bfs.popleft()
        if j == m - 1:
            # Calculate the minimum time to reach destination (n-1, m-1)
            # considering all possible positions the robot can be in at time t
            ans = t + min((i - t) % n + 1, n - 1 - (i - t) % n)
            break
        # Move right: check for rock at (i+1,j+1) which will be at (i+1,j+1) at time t
        if g[(i + 1) % n][(j + 1) % m] == 0 and not vis[(i + 1) % n][(j + 1) % m]:
            vis[(i + 1) % n][(j + 1) % m] = 1
            bfs += [((i + 1) % n, (j + 1) % m, t + 1)]
        # Move down: check for rock at (i+1,j) and (i+2,j) which will be at (i+1,j) and (i+2,j) at time t
        if g[(i + 1) % n][j] == g[(i + 2) % n][j] == 0 and not vis[(i + 2) % n][j]:
            vis[(i + 2) % n][j] = 1
            bfs += [((i + 2) % n, j, t + 1)]

    print(ans if ans < inf else -1)


# https://github.com/VaHiX/CodeForces/

File: CPP/ByTier/F/1933_F_Turtle_Mission_Robot_and_the_Earthquake.cpp
// Problem: CF 1933 F - Turtle Mission: Robot and the Earthquake
// https://codeforces.com/contest/1933/problem/F

/*
Algorithm: BFS with time-aware state tracking
Approach: 
- Use BFS to explore all possible paths from (0,0) to (n-1,m-1)
- State representation: (x,y) coordinates are packed into a single integer
- The grid is cyclic in the vertical direction, so we need to track the movement of rocks over time
- BFS is performed with time tracking, and we check for valid moves at each time step
- Optimization: use two queues for bidirectional BFS, and a packed coordinate system for efficient memory usage

Time Complexity: O(n * m * (n + m))
Space Complexity: O(n * m)

Key techniques:
1. Bit manipulation for packed coordinates: x and y are combined into one integer
2. Two queue BFS to reduce time complexity
3. In-place rock state updating to avoid extra memory
*/
#include <iostream>
#include <vector>
#include <algorithm>

const int shift = 10;
const int mask = (1 << shift) - 1;
int n, m;
inline int X(int xy) { return xy & mask; } // Extract x coordinate from packed integer
inline int Y(int xy) { return xy >> shift; } // Extract y coordinate from packed integer
inline int xy(int x, int y) { return (y << shift) + (x % n); } // Pack x,y coordinates into integer
char a[1 << 2 * shift]; // Grid state storage using packed coordinates
std::vector<int> q[2]; // Two queues for BFS with alternating usage
void proc() {
  std::cin >> n >> m;
  for (int i = 0; i < n; i++)
    for (int j = 0; j < m; j++)
      std::cin >> a[xy(i, j)]; // Read in the grid
  q[0].clear();
  q[1].clear();
  q[0].push_back(xy(0, 0)); // Start from (0,0)
  a[xy(0, 0)] = '2'; // Mark as visited
  const int mx = 1'000'000'000;
  int r = mx; // Result variable to store minimum time
  for (int t = 0; t < r && q[t % 2].size(); t++) { // BFS with time tracking
    for (auto v : q[t % 2]) {
      int x = X(v), y = Y(v); // Unpack current state
      if (y == m - 1) // If at the destination column
        r = std::min(r, ((x + n) - (n - 1 + t) % n) % n + t); // Calculate time to reach destination
      // Right move: check if next cell is safe and not occupied
      if (y < m - 1 && a[xy(x + 1, y + 1)] == '0') {
        q[1 - t % 2].push_back(xy(x + 1, y + 1)); // Add to next queue
        a[xy(x + 1, y + 1)] = '2'; // Mark as visited
      }
      // Down move: check if next cell is safe and not occupied by falling rock
      if (a[xy(x + 1, y)] != '1' && a[xy(x + 2, y)] == '0') {
        q[1 - t % 2].push_back(xy(x + 2, y)); // Add to next queue
        a[xy(x + 2, y)] = '2'; // Mark as visited
      }
    }
    q[t % 2].clear(); // Clear current queue for next iteration
  }
  std::cout << (r == mx ? -1 : r) << std::endl; // Output result or -1 if impossible
}
int main() {
  q[0].reserve(1'000'000); // Pre-allocate memory for queues
  q[1].reserve(1'000'000);
  std::cin.tie(0);
  std::cin.sync_with_stdio(false);
  std::cout.sync_with_stdio(false);
  int t = 1;
  std::cin >> t;
  while (t--)
    proc();
  return 0;
}


// https://github.com/VaHiX/CodeForces/